firestore/bulkwriter.go

```diff
@@ -0,0 +1,325 @@
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newUpdatePathWrites(updates, preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot update doc %v", doc.ID)
+	}
+
+	if len(w) > 1 {
+		return nil, fmt.Errorf("firestore: too many writes sent to bulkwriter")
+	}
+
+	j := bw.write(w[0])
+	return j, nil
+}
+
+// checkConditions determines whether this write attempt is valid. It returns
+// an error if either the BulkWriter has already been closed or if it
+// receives a nil document reference.
+func (bw *BulkWriter) checkWriteConditions(doc *DocumentRef) error {
+	bw.openLock.Lock()
+	open := bw.isOpen
+	bw.openLock.Unlock()
+	if !open {
+		return errors.New("firestore: BulkWriter has been closed")
+	}
+
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, havePath := bw.docUpdatePaths[doc.shortPath]
+	if havePath {
+		return fmt.Errorf("firestore: BulkWriter received duplicate write for path: %v", doc.shortPath)
+	}
+
+	bw.docUpdatePaths[doc.shortPath] = true
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) *BulkWriterJob {
+
+	j := &BulkWriterJob{
+		resultChan: make(chan bulkWriterResult, 1),
+		write:      w,
+		ctx:        bw.ctx,
+	}
+
+	bw.limiter.Wait(bw.ctx)
+	// ignore operation size constraints and related errors; can't be inferred at compile time
+	// Bundler is set to accept an unlimited amount of bytes
+	_ = bw.bundler.Add(j, 0)
+
+	return j
+}
+
+// send transmits writes to the service and matches response results to job channels.
+func (bw *BulkWriter) send(i interface{}) {
+	bwj := i.([]*BulkWriterJob)
+
+	if len(bwj) == 0 {
+		return
+	}
+
+	var ws []*pb.Write
+	for _, w := range bwj {
+		ws = append(ws, w.write)
+	}
+
+	bwr := &pb.BatchWriteRequest{
+		Database: bw.database,
+		Writes:   ws,
+		Labels:   map[string]string{},
+	}
+
+	select {
+	case <-bw.ctx.Done():
+		return
+	default:
+		resp, err := bw.vc.BatchWrite(bw.ctx, bwr)
+		if err != nil {
+			// Do we need to be selective about what kind of errors we send?
+			for _, j := range bwj {
+				j.setError(err)
+			}
+			return
+		}
+		// Match write results with BulkWriterJob objects
+		for i, res := range resp.WriteResults {
+			s := resp.Status[i]
+			c := s.GetCode()
+			if c != 0 { // Should we do an explicit check against rpc.Code enum?
+				j := bwj[i]
+				j.attempts++
+
+				// Do we need separate retry bundler?
+				if j.attempts < maxRetryAttempts {
+					// ignore operation size constraints and related errors; job size can't be inferred at compile time
+					// Bundler is set to accept an unlimited amount of bytes
+					_ = bw.bundler.Add(j, 0)
+				} else {
+					j.setError(fmt.Errorf("firestore: write failed with status: %v", s))
+				}
+				continue
+			}
+
+			bwj[i].resultChan <- bulkWriterResult{err: nil, result: res}
```

Comment:
nit: Either approach seems fine, but this should probably match `setError` above, or vice versa.

Reply:
IIRC, @codyoss suggested getting rid of `setResult()` since it's only used once.